### PR TITLE
kv: disable circuit breaker on destroyed replica

### PR DESCRIPTION
**Pull request by lyang24**
This commit disables circuit breaker on replica that is destoryed or in the process of being destroyed.

Informs #104567

Release note: None

### CODE REVIEW

**Review by lyang24 — COMMENTED** [review]
I wonder what is the best way to best this. I see majority of the circuit breaker tests are in `pkg/kv/kvserver/client_replica_circuit_breaker_test.go`. However, that file is under a different go pkg and replica's destroyed status is not accessible from the testfile's package. my testing idea is something like below.


```
func (cbt *circuitBreakerTest) SetReplicaAsDestroyed(idx int) {
	r := cbt.repls[idx].Replica
	r.mu.Lock()
	defer r.mu.Unlock()
	r.mu.destroyStatus.Set(errors.New("test"), destroyReasonRemoved)
}

func TestReplicaCircuitBreaker_DestroyedReplicaDisabled(t *testing.T) {
	defer leaktest.AfterTest(t)()
	defer log.Scope(t).Close(t)
	tc := setupCircuitBreakerTest(t)
	defer tc.Stopper().Stop(context.Background())

	// Put the lease on n1 but then trip the breaker with the probe
	// disabled.
	require.NoError(t, tc.Write(n1))
	tc.SetProbeEnabled(n1, false)
	tc.TripBreaker(n1)
	// Set n1 as destroyed
	tc.SetReplicaAsDestroyed(n1)
	resumeHeartbeats := tc.ExpireAllLeasesAndN1LivenessRecord(t, pauseHeartbeats)


	resumeHeartbeats()
	testutils.SucceedsSoon(t, func() error {
		err := tc.Read(n1)
		if errors.HasType(err, (*kvpb.NotLeaseHolderError)(nil)) {
			return err
		}
		require.NoError(t, err)
		return nil
	})
}
```

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119719#-:-NrnW8CT2-YY7VIPFqF7:b-l3qi60)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker and @pav-kv)

**Review by lyang24 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119719)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker and @pav-kv)

___
*[`pkg/kv/kvserver/replica_circuit_breaker.go` line 86 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/119719#-NrnYKBYEeJAvgaS4S7Y:-NrnYKBYEeJAvgaS4S7Z:bf9z1fo) ([raw file](https://github.com/cockroachdb/cockroach/blob/6c79d3ccda41b25c0b769619ac7ab405414f6bd6/pkg/kv/kvserver/replica_circuit_breaker.go#L86)):*
> ```Go
> func (br *replicaCircuitBreaker) enabled() bool {
> 	// Circuit break will not be enabled on destroyed replica.
> 	d, _ := br.r.IsDestroyed()
> ```

the other approach i had in mind is to move this logic to the ontrip function and left a `circuit breaker is supressed due to destroyed replica %s, repl.Desc()` warning

**Review by lyang24 — COMMENTED** [review]
looked a bit into failed tests, the current approach is failing because in situations that replica.mu is locked by the process such as `refreshProposalsLocked` and then circuit breaker tripped and trying to determine if is enabled as the first task. the new introduced process will read from replica for destoryed status and unfortunately that need read lock.

I wonder if i should just pivot back to add the check in this line
https://github.com/lyang24/cockroach/blob/2f7bf071af146e1e238f985884b99b2a6636115a/pkg/kv/kvserver/replica_raft.go#L1467

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119719#-:-NrsP78RC5mD17T047gC:b-7qrg12)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker and @pav-kv)

**Review by erikgrinaker — COMMENTED** [review]
I think there's two parts to this: we shouldn't let a circuit breaker trip if the replica has already been destroyed, and we should untrip the breaker when the replica gets destroyed (in particular, we must stop the probe to avoid goroutine leaks).

One place to start might be to return early in `sendProbe` if the replica is destroyed, which would immediately untrip the breaker. We already have a similar condition there for uninitialized replicas:

https://github.com/cockroachdb/cockroach/blob/7e4233d09edef5420b8f12d917112ce762ac1855/pkg/kv/kvserver/replica_circuit_breaker.go#L212-L214

This should avoid the deadlock, because this runs async and will wait for the current caller to release the lock. I'm not sure how this affects an already in-flight proposal, because I think we'll be waiting for it indefinitely, but if we cancel those when the replica is destroyed then this might be sufficient.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119719#-:-NsJEfI48RRU1bDLGvEs:b-3jsn9q)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @pav-kv)

File: pkg/kv/kvserver/replica_circuit_breaker.go [hunk 1]
```diff
@@ -209,7 +210,8 @@ func sendProbe(ctx context.Context, r replicaInCircuitBreaker) error {
 	// enhance the probe, we may need to allow any additional requests we send to
 	// chose to bypass the circuit breaker explicitly.
 	desc := r.Desc()
-	if !desc.IsInitialized() {
+	destroyed, _ := r.IsDestroyed()
+	if !desc.IsInitialized() || destroyed == destroyReasonRemoved {
```
Comment (pav-kv) [inline pkg/kv/kvserver/replica_circuit_breaker.go:214]: Please add a comment here, something like:

```golang
// Untrip the breaker if the replica is destroyed or not initialized.
```

File: pkg/kv/kvserver/replica_circuit_breaker.go [hunk 2]
```diff
@@ -209,7 +210,8 @@ func sendProbe(ctx context.Context, r replicaInCircuitBreaker) error {
...
 		return nil
 	}
```
Comment (pav-kv) [inline pkg/kv/kvserver/replica_circuit_breaker.go:216]:
```suggestion
	if reason, _ := r.IsDestroyed(); !desc.IsInitialized() || reason.Removed() {
		return nil
	}
```

File: pkg/kv/kvserver/replica_raft.go [hunk 3]
```diff
@@ -1449,9 +1449,11 @@ func (r *Replica) refreshProposalsLocked(
 	}
 
 	r.mu.slowProposalCount = slowProposalCount
+	destroyed := r.mu.destroyStatus.Removed()
 
 	// If the breaker isn't tripped yet but we've detected commands that have
-	// taken too long to replicate, trip the breaker now.
+	// taken too long to replicate and replica do not have destroyed status,
```
Comment (pav-kv) [inline pkg/kv/kvserver/replica_raft.go:1455]:
```suggestion
	// taken too long to replicate, and the replica is not destroyed,
```

File: pkg/kv/kvserver/replica_raft.go [hunk 4]
```diff
@@ -1464,7 +1466,7 @@ func (r *Replica) refreshProposalsLocked(
 	// already tripped and no probe is running, thus ensuring that even if a
 	// request got added in while the probe was about to shut down, there will
 	// be regular attempts at healing the breaker.
-	if maxSlowProposalDuration > 0 && r.breaker.Signal().Err() == nil {
+	if maxSlowProposalDuration > 0 && r.breaker.Signal().Err() == nil && !destroyed {
```
Comment (pav-kv) [inline pkg/kv/kvserver/replica_raft.go:1469]: Is it possible to be in this branch on a destroyed replica?

If the replica is destroyed, can we have a spinning behaviour here?
1. `r.breaker.Signal().Err() == nil` leads to tripping the breaker with `r.breaker.TripAsync(err)`
2. `sendProbe` (modified above in this PR) will untrip the breaker
3. repeat

If we can, how do we avoid it?

Comment (pav-kv) [reply pkg/kv/kvserver/replica_raft.go:1469]: Ah, we won't enter this loop if `destroyed` is true.

**Review by pav-kv — APPROVED** [review]

**Review by lyang24 — COMMENTED** [review]
TFTR

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119719#-:-Nu-XycC9yehKdRwe-SC:bb73zsl)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker and @pav-kv)

___
*[`pkg/kv/kvserver/replica_circuit_breaker.go` line 216 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119719#gh-1521184252:-Nu-Xl8f-tFsN6q-GvJd:b12nqam) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ce715ece22b35259ceef217c8d58abe9f3e68cc/pkg/kv/kvserver/replica_circuit_breaker.go#L216)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

```suggestion
	if reason, _ := r.IsDestroyed(); !desc.IsInitialized() || reason.Removed() {
		return nil
	}
```
</blockquote></details>

Removed() is not available for DestroyReason struct I adjusted to `reason != destroyReasonRemoved` instead
___
*[`pkg/kv/kvserver/replica_raft.go` line 1455 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119719#gh-1521185882:-Nu-XxaMEs15SFRHmx8o:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ce715ece22b35259ceef217c8d58abe9f3e68cc/pkg/kv/kvserver/replica_raft.go#L1455)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

```suggestion
	// taken too long to replicate, and the replica is not destroyed,
```
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/replica_raft.go` line 1469 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119719#gh-1521196008:-Nu-XyAL39bMiaTT7Vxw:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ce715ece22b35259ceef217c8d58abe9f3e68cc/pkg/kv/kvserver/replica_raft.go#L1469)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Ah, we won't enter this loop if `destroyed` is true.
</blockquote></details>

Done.

File: pkg/kv/kvserver/replica_circuit_breaker.go [hunk 5]
```diff
@@ -209,7 +210,8 @@ func sendProbe(ctx context.Context, r replicaInCircuitBreaker) error {
 	// enhance the probe, we may need to allow any additional requests we send to
 	// chose to bypass the circuit breaker explicitly.
 	desc := r.Desc()
-	if !desc.IsInitialized() {
+	// Untrip the breaker if the replica is destroyed or not initialized.
+	if reason, _ := r.IsDestroyed(); !desc.IsInitialized() || reason != destroyReasonRemoved {
```
Comment (pav-kv) [inline pkg/kv/kvserver/replica_circuit_breaker.go:214]: should this be `reason == destroyReasonRemoved`?

**Review by lyang24 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/119719)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @erikgrinaker and @pav-kv)

___
*[`pkg/kv/kvserver/replica_circuit_breaker.go` line 214 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/119719#gh-1521180385:-Nu-b_yb9d45QFUFTfxZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6ce715ece22b35259ceef217c8d58abe9f3e68cc/pkg/kv/kvserver/replica_circuit_breaker.go#L214)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

Please add a comment here, something like:

```golang
// Untrip the breaker if the replica is destroyed or not initialized.
```
</blockquote></details>

Done.
___
*[`pkg/kv/kvserver/replica_circuit_breaker.go` line 214 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/119719#gh-1541372555:-Nu-ba_57V-R06vWOF1P:b-oa155d) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe481f12076132fc61c1b16c86d12717f31a227e/pkg/kv/kvserver/replica_circuit_breaker.go#L214)):*
<details><summary><i>Previously, pav-kv (Pavel Kalinnikov) wrote…</i></summary><blockquote>

should this be `reason == destroyReasonRemoved`?
</blockquote></details>

yes my bad

**Pull request merged**